contracts/TwabRewards.sol

```diff
@@ -0,0 +1,451 @@
+        returns (Epoch memory)
+    {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        return _getEpoch(_epochId, _promotion);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRewardAmount(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) external view override returns (uint256) {
+        return _calculateRewardAmount(_user, _promotionId, _epochId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function claimRewards(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) external override returns (uint256) {
+        require(_epochId < _getCurrentEpoch(_promotionId).id, "TwabRewards/epoch-not-over");
+        require(
+            !_isClaimedEpoch(_user, _promotionId, _epochId),
+            "TwabRewards/rewards-already-claimed"
+        );
+
+        uint256 _rewardAmount = _calculateRewardAmount(_user, _promotionId, _epochId);
+
+        IERC20 _token = IERC20(_getPromotion(_promotionId).token);
+        _token.safeTransferFrom(address(this), _user, _rewardAmount);
+
+        _setClaimedEpoch(_claimedEpochs[_promotionId][_user], _epochId, true);
+
+        emit RewardsClaimed(_token, _rewardAmount);
+
+        return _rewardAmount;
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+    @notice Determine if address passed is actually a ticket.
+    @param _ticket Address to check
+   */
+    function _requireTicket(address _ticket) internal view {
+        require(address(_ticket) != address(0), "TwabRewards/ticket-not-zero-address");
+
+        (bool succeeded, bytes memory data) = address(_ticket).staticcall(
+            abi.encodePacked(ITicket(_ticket).controller.selector)
+        );
+
+        address controllerAddress;
+
+        if (data.length > 0) {
+            controllerAddress = abi.decode(data, (address));
+        }
+
+        require(succeeded && controllerAddress != address(0), "TwabRewards/invalid-ticket");
+    }
+
+    /**
+        @notice Determine if the number of epochs passed exceeds the maximum number of epochs.
+        @param _numberOfEpochs Number of epochs to check
+    */
+    function _requireEpochLimit(uint256 _numberOfEpochs) internal pure {
+        require(_numberOfEpochs < 256, "TwabRewards/exceeds-256-epochs-limit");
+    }
+
+    /**
+        @notice Get settings for a specific promotion.
+        @dev Will revert if the promotion does not exist.
+        @param _promotionId Promotion id to get settings for
+        @return Promotion settings
+     */
+    function _getPromotion(uint256 _promotionId) internal view returns (Promotion memory) {
+        return _promotions[_promotionId];
+    }
+
+    /**
+        @notice Get settings for a specific epoch.
+        @param _epochId Epoch id to get settings for
+        @param _promotion Promotion settings
+        @return Epoch settings
+     */
+    function _getEpoch(uint256 _epochId, Promotion memory _promotion)
```

Comment:
This is a PERFECT example of brittle code

You're passing a struct, but the function only uses the epochDuration and startTimestamp.  Just pass the epoch duration and start timestamp!

By having the user pass a whole struct, this function brittle to change in the struct.
The user now also needs to pass an entire struct just to compute an epoch id.

this function should prob just be refactored into `_getEpochStartTime`

Reply:
I've removed this function in favor of `_getCurrentEpochId` since we can calculate the epoch start time by doing: `uint256 _epochStartTimestamp = _promotion.startTimestamp + (_epochDuration * _epochId);`